File: bumi/src/http/inputs.ts
import type { MatchingData } from "../types.ts";
import type { HttpRequest } from "./request.ts";

/** Extract the path (without query string) from an HTTP request. */
export class PathInput {
	get(ctx: HttpRequest): MatchingData {
		return ctx.path;
	}
}

/** Extract the HTTP method from a request. */
export class MethodInput {
	get(ctx: HttpRequest): MatchingData {
		return ctx.method;
	}
}

/** Extract a header value by name (case-insensitive). */
export class HeaderInput {
	constructor(readonly name: string) {}

	get(ctx: HttpRequest): MatchingData {
		return ctx.header(this.name);
	}
}

/** Extract a query parameter value by name. */
export class QueryParamInput {
	constructor(readonly name: string) {}

	get(ctx: HttpRequest): MatchingData {
		return ctx.queryParam(this.name);
	}
}

/**
 * Extract the `:authority` pseudo-header (Host in HTTP/1).
 *
 * Read from the headers, which is where rumi's `HttpMessage` reads it from
 * too. Without this and `SchemeInput`, four of the six `xuma.http.v1.*` type
 * URLs resolved here and six resolved in rumi — a type URL that loads in one
 * implementation and not another.
 */
export class AuthorityInput {
	get(ctx: HttpRequest): MatchingData {
		return ctx.header(":authority");
	}
}

/** Extract the `:scheme` pseudo-header (http/https). */
export class SchemeInput {
	get(ctx: HttpRequest): MatchingData {
		return ctx.header(":scheme");
	}
}

File: puma/src/xuma/http/_inputs.py
"""DataInput implementations for HttpRequest.

Each input extracts a specific field from an HttpRequest context
and returns it as a MatchingData for predicate evaluation.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from xuma._types import MatchingData
    from xuma.http._request import HttpRequest


@dataclass(frozen=True, slots=True)
class PathInput:
    """Extracts the request path (without query string)."""

    def get(self, ctx: HttpRequest, /) -> MatchingData:
        return ctx.path


@dataclass(frozen=True, slots=True)
class MethodInput:
    """Extracts the HTTP method (case-sensitive)."""

    def get(self, ctx: HttpRequest, /) -> MatchingData:
        return ctx.method


@dataclass(frozen=True, slots=True)
class HeaderInput:
    """Extracts a header value by name (case-insensitive lookup)."""

    name: str

    def get(self, ctx: HttpRequest, /) -> MatchingData:
        return ctx.header(self.name)


@dataclass(frozen=True, slots=True)
class QueryParamInput:
    """Extracts a query parameter value by name."""

    name: str

    def get(self, ctx: HttpRequest, /) -> MatchingData:
        return ctx.query_param(self.name)


@dataclass(frozen=True, slots=True)
class AuthorityInput:
    """Extracts the ``:authority`` pseudo-header (Host in HTTP/1).

    Read from the headers, which is where rumi's `HttpMessage` reads it from
    too. Without this and `SchemeInput`, four of the six ``xuma.http.v1.*``
    type URLs resolved here and six resolved in rumi — a type URL that loads in
    one implementation and not another.
    """

    def get(self, ctx: HttpRequest, /) -> MatchingData:
        return ctx.header(":authority")


@dataclass(frozen=True, slots=True)
class SchemeInput:
    """Extracts the ``:scheme`` pseudo-header (http/https)."""

    def get(self, ctx: HttpRequest, /) -> MatchingData:
        return ctx.header(":scheme")
